ttnn/cpp/ttnn/operations/ccl/all_gather_matmul/device/multi_core/all_gather_matmul_op_multi_core.cpp

```diff
@@ -0,0 +1,167 @@
+    );
+
+
+    // Setup cores used for datacopy
+    std::vector<CoreCoord> all_datacopy_cores;
+    all_datacopy_cores.reserve(1);
+
+
+    // Select cores for datacopy (single core for now)
+    CoreRangeSet datacopy_workers = CoreRangeSet({CoreRange(datacopy_core_coord)});
+
+    auto datacopy_cores = corerange_to_cores(datacopy_workers, std::nullopt, true);
+    all_datacopy_cores.insert(all_datacopy_cores.end(), datacopy_cores.begin(), datacopy_cores.end());
+
+    std::cout << "Finished setting up datacopy cores." << std::endl;
+
+    // Setup semaphores used to signal datacopy. TODO: instead of datacopy, this should be matmul cores
+    auto datacopy_signal_semaphore_addr_dir0 = CreateSemaphore(program, datacopy_workers, 0);
+    auto datacopy_signal_semaphore_addr_dir1 = CreateSemaphore(program, datacopy_workers, 0);
+
+    // Setup args for the kernel
+    const uint32_t tile_size = 32;
+    const uint32_t page_size = all_gather_output_tensor.buffer()->page_size();
+
+    const tt::DataFormat cb_data_format = tt::tt_metal::datatype_to_dataformat_converter(all_gather_output_tensor.get_dtype());
+
+
+    auto all_gather_output_buffer = all_gather_output_tensor.buffer();
+    auto datacopy_output_buffer = datacopy_output_tensor.buffer();
+
+    bool all_gather_output_is_dram = all_gather_output_buffer->buffer_type() == tt::tt_metal::BufferType::DRAM ? 1 : 0;
+    bool datacopy_output_is_dram = datacopy_output_buffer->buffer_type() == tt::tt_metal::BufferType::DRAM ? 1 : 0;
+
+    uint32_t last_output_page_offset = (ring_size - 1) * tensor_slicer.output_page_offset;
+
+    uint32_t num_rows = input_tensor.get_legacy_shape()[2] / tile_size ;
+
+
+    // Compile time args
+    std::vector<uint32_t> datacopy_ct_args = {
+        static_cast<uint32_t>(all_gather_output_is_dram),
+        static_cast<uint32_t>(datacopy_output_is_dram),
+        static_cast<uint32_t>(num_transfers),
+        static_cast<uint32_t>(page_size),
+        static_cast<uint32_t>(ring_index),
+        static_cast<uint32_t>(ring_size),
+        static_cast<uint32_t>(all_gather_output_tensor.get_legacy_shape()[3] / tile_size), // tesnor width
+        static_cast<uint32_t>(all_gather_output_tensor.get_legacy_shape()[2] / tile_size), // tensor height
+        static_cast<uint32_t>(tensor_slicer.num_cols), // tensor slice width in tiles
+        static_cast<uint32_t>(num_rows), // tnesor slice height in tiles
+        static_cast<uint32_t>(tensor_slicer.output_page_offset),
+        static_cast<uint32_t>(last_output_page_offset),
+        static_cast<uint32_t>(true ? 1 : 0), // TODO
+        static_cast<uint32_t>(datacopy_signal_semaphore_addr_dir0),
+        static_cast<uint32_t>(datacopy_signal_semaphore_addr_dir1),
+    };
+
+    uint32_t cb_id_in0 = tt::CB::c_in0;
+    tt::tt_metal::CircularBufferConfig cb_in0_config =
+        tt::tt_metal::CircularBufferConfig(
+            page_size * 200 /* TODO: Update to be actual number */, {{cb_id_in0, cb_data_format}})
```

Comment:
Same comment about hardcoded page-count. Please move to named var and pass around that way. Will make updating in the future easier too.